src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,776 @@
+﻿using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    /// <summary>
+    /// Represents a C# or VB expression.
+    /// </summary>
+    public interface IExpression : IOperation
+    {
+        /// <summary>
+        /// Result type of the expression.
+        /// </summary>
+        ITypeSymbol ResultType { get; }
+        /// <summary>
+        /// If the expression evaluates to a constant value, the value of the expression, and otherwise null.
+        /// </summary>
+        object ConstantValue { get; }
```

Comment:
The ConstantValue type is not presently public. If we want to expose that type, using it here would certainly make sense. The ConstantValue type is designed for use within a compiler, including efficient encodings for common values, but its API is probably sufficiently approachable. "object" looks mysterious, but the usage pattern is straightforward:

    if (mumble.ConstantValue is int)
    {
        var intVal = (int)mumble.ConstantValue;
    }

All of this is a long-winded way of saying "I dunno". Opinions?